A animation helper that morphs between two controls

HHChaos opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- 👉 It is imperative to resolve ONE ISSUE PER PR and avoid making multiple changes unless the changes interrelate with each other -->

<!-- 📝 Please always keep the "☑️ Allow edits by maintainers" button checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork. This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->

## Feature #4323

A new  animation helper that morphs between two controls was added in this PR and an example for it.


![animation](https://user-images.githubusercontent.com/9880072/138796652-7b60bc81-6428-425a-832d-9a8b5e1d1253.gif)
 
<!-- Add the relevant issue number after the word "Fixes" mentioned above (for ex: `## Fixes #0000`) which will automatically close the issue once the PR is merged. -->

<!-- Add a brief overview here of the feature/bug & fix. -->

## PR Type

What kind of change does this PR introduce?

<!-- Please uncomment one or more options below that apply to this PR. -->

<!-- - Bugfix -->
- Feature
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Sample app changes -->
<!-- - Other... Please describe: -->

## PR Checklist

Please check if your PR fulfills the following requirements: <!-- and remove the ones that are not applicable to the current PR -->

- [x] Tested code with current [supported SDKs](../#supported)
- [x] New component
  - [ ] Pull Request has been submitted to the documentation repository [instructions](../blob/main/Contributing.md#docs). Link: <!-- docs PR link -->
  - [ ] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
  - [ ] If control, added to Visual Studio Design project
- [x] Sample in sample app has been added / updated (for bug fixes / features)
  - [x] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/CommunityToolkit/WindowsCommunityToolkit-design-assets)
- [ ] New major technical changes in the toolkit have or will be added to the [Wiki](https://github.com/CommunityToolkit/WindowsCommunityToolkit/wiki) e.g. build changes, source generators, testing infrastructure, sample creation changes, etc...
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Header has been added to all new source files (run _build/UpdateHeaders.bat_)
- [ ] Contains **NO** breaking changes

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->

## Other information
The document is not ready.
<!-- Please add any other information that might be helpful to reviewers. -->

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Animations/Helpers/TransitionHelper.Properties.cs:
```diff
@@ -0,0 +1,120 @@
+        /// <summary>
+        /// Gets or sets the collection of animation configurations of UI elements that need to be connected by animation.
+        /// </summary>
+        public List<AnimationConfig> AnimationConfigs
```
If none of these child elements will contain bindings, you should be able to set a ContentPropertyAttribute on the class to specify this property, then your XAML won't need to call out the property and can be just this:
```xml
<ani:TransitionHelper x:Key="MyTransitionHelper">
    <ani:AnimationConfig Id="background" ScaleMode="Scale" />
    <ani:AnimationConfig Id="image" ScaleMode="Scale" />
    <ani:AnimationConfig Id="guide" />
    <ani:AnimationConfig Id="name" ScaleMode="ScaleY" />
    <ani:AnimationConfig Id="desc" />
</ani:TransitionHelper>
```

Otherwise if the child elements would need bindings, we'd probably have to inherit from DependencyObjectCollection like we needed for `AnimationSet` and `AnimationScope`.

Eh, @Sergio0694?

HHChaos replied:
I added `ContentProperty` attribute for `TransitionHelper`, I think it should not need bindings, `ContentProperty` is ok for it.

HHChaos commented:
@michael-hawker Do I need to add separate samples for `StartTransitionAction` and `ReverseTransitionAction`? Actually I think this sample seems to be enough.

Arlodotexe commented on Microsoft.Toolkit.Uwp.UI.Animations/Helpers/TransitionHelper.AttachedProperty.cs:
```diff
@@ -0,0 +1,94 @@
+                return null;
+            }
+
+            var elements = targetElement.FindDescendants().ToList() ?? new List<DependencyObject>();
```
@HHChaos These should be available to use now 🙂

HHChaos replied:
yean, done

Arlodotexe commented:
> @michael-hawker Do I need to add separate samples for `StartTransitionAction` and `ReverseTransitionAction`? Actually I think this sample seems to be enough.

It would be useful to first-timers if we had a simple example alongside the more complex one.

Arlodotexe reviewed (commented):
Looks pretty good, had a few cleanup suggestions and questions. 

One more general question to consider. While using Composition by default is obviously a good choice, there may be scenarios where using the XAML layer is required. Since the AnimationBuilder APIs make this easy, should we expose an option to use either Composition or XAML layer?

Sergio0694 commented on Microsoft.Toolkit.Uwp.UI.Animations/Helpers/TransitionHelper.AttachedProperty.cs:
```diff
@@ -0,0 +1,94 @@
+            var elements = targetElement.FindDescendants().ToList() ?? new List<DependencyObject>();
+            elements.Add(targetElement);
+
+            return elements.Where(element => GetIsIgnored(element)).OfType<UIElement>();
```
Note: that results in worse performance due to Roslyn not caching delegates created using the method group syntax. It's fixed in main now but won't be available until C# 11. I'd say it's fine to leave this as is for now.

HHChaos replied:
Thanks @Arlodotexe and @Sergio0694 , keep it for now

HHChaos commented on Microsoft.Toolkit.Uwp.UI.Animations/Enums/ScaleMode.cs:
```diff
@@ -0,0 +1,34 @@
+    /// <summary>
+    /// Indicates the strategy when the scale property of a UI element is animated.
+    /// </summary>
+    public enum ScaleMode
```
Actually here ScaleX/ScaleY doesn't mean it is only for horizontal/vertical scaling, It's just a scaling strategy. So using flag might not be appropriate

HHChaos commented on Microsoft.Toolkit.Uwp.UI.Animations/Helpers/TransitionHelper.Logic.cs:
```diff
@@ -0,0 +1,327 @@
+        {
+            this.ToggleVisualState(this.Source, this.SourceToggleMethod, true);
+            this.ToggleVisualState(this.Target, this.TargetToggleMethod, true);
+            if (this._needUpdateTargetLayout && this.TargetToggleMethod == VisualStateToggleMethod.ByVisibility)
```
@Arlodotexe I updated the logic here, I believe the ambiguity here is resolved

Arlodotexe replied:
Nicely cleaned up 🙂

HHChaos commented:
> > @michael-hawker Do I need to add separate samples for `StartTransitionAction` and `ReverseTransitionAction`? Actually I think this sample seems to be enough.
> 
> @HHChaos It would be useful to first-timers if we had a simple example alongside the more complex one.

I have added a new example for them.

HHChaos commented:
> Looks pretty good, had a few cleanup suggestions and questions.
> 
> One more general question to consider. While using Composition by default is obviously a good choice, there may be scenarios where using the XAML layer is required. Since the AnimationBuilder APIs make this easy, should we expose an option to use either Composition or XAML layer?

I can't think of such a scenario, and based on performance considerations, I think it's ok now.🤔

michael-hawker commented:
If everything under the hood is just using the Toolkit Animation APIs, it probably wouldn't hurt to put a property for the Animation layer on the TransitionHelper then. @Sergio0694 thoughts/concerns with that? Would be handy to get you on the PR review for this too. 🙂

michael-hawker commented:
Hey @HHChaos got to play with this a little more today. **_This is super cool._** 🦙❤

Probably something I'm going to suggest we bring into Labs (see #4487) to get into folks hands sooner as 8.0 is a ways away. (Though we'll have to figure out some dependency things for that maybe.)

Anyway, I did notice a couple of minor things from a sample/interaction perspective:

1) The in `Light` mode (at least on Windows 10) you can't really see the border, so it's hard to know that's animating as well:

![image](https://user-images.githubusercontent.com/24302614/168396637-6d68c4d3-c176-4b7f-bcd7-19d950bc7f45.png)

Compared to dark mode:

![image](https://user-images.githubusercontent.com/24302614/168396660-fbc0770f-1d7e-4a63-904d-f270a68939f1.png)

2) I'm noticing the tiniest of 'flicker' when the animation starts. I hypothesize this is from the switch in visibilities between the source and the target elements at the start of the animation.

i.e. I think we need to ensure that the element which is currently invisible becomes visible first before the element which is visible becomes collapsed (regardless of the direction of the animation)

I'm not entirely sure if this is based on the visibility toggle ordering or something based with the Opacity manipulation that is occurring also?

Thoughts?

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Behaviors/Animations/StartTransitionAction.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <inheritdoc/>
+        public object Execute(object sender, object parameter)
+        {
+            if (this.Transition is null)
+            {
+                throw new ArgumentNullException(nameof(this.Transition));
+            }
```
@Sergio0694 do you have an internal throw helper in the animations package already we can just leverage?

HHChaos commented:
> Hey @HHChaos got to play with this a little more today. **_This is super cool._** 🦙❤
> 
> Probably something I'm going to suggest we bring into Labs (see #4487) to get into folks hands sooner as 8.0 is a ways away. (Though we'll have to figure out some dependency things for that maybe.)
> 
> Anyway, I did notice a couple of minor things from a sample/interaction perspective:
> 
> 1. In `Light` mode (at least on Windows 10) you can't really see the border, so it's hard to know that's animating as well:
> 
> ![image](https://user-images.githubusercontent.com/24302614/168396637-6d68c4d3-c176-4b7f-bcd7-19d950bc7f45.png)
> 
> Compared to dark mode:
> 
> ![image](https://user-images.githubusercontent.com/24302614/168396660-fbc0770f-1d7e-4a63-904d-f270a68939f1.png)
> 
> 2. I'm noticing the tiniest of 'flicker' when the animation starts. I hypothesize this is from the switch in visibilities between the source and the target elements at the start of the animation.
> 
> i.e. I think we need to ensure that the element which is currently invisible becomes visible first before the element which is visible becomes collapsed (regardless of the direction of the animation)
> 
> I'm not entirely sure if this is based on the visibility toggle ordering or something based with the Opacity manipulation that is occurring also?
> 
> Thoughts?

Hi @michael-hawker ,
For 1st issue, this seems to be a problem with `AttachedCardShadow`? I can't reproduce this problem in Win11. Anyway, I removed the `AttachedCardShadow`, and changed the background to fix it.
For 2nd issue,  it is actually a complex issue, since our opacity animation doesn't change the order of elements on the Z axis, so it's not just a visible/invisible state toggle, transition animations may also be affected by elements with higher Z-Index, I've noticed this problem before, but I didn't find a perfect solution. I've added a `TransitionMode` to help mitigate this issue, you can check if that works.

michael-hawker commented:
@HHChaos ah, you can't apply an `AttachedShadow` directly to a border, see the example here:

https://github.com/CommunityToolkit/WindowsCommunityToolkit/blob/2d9041626d9dd7c90be94558b166e378ac2eb845/Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Shadows/AttachedShadowWin2DXaml.bind#L44-L61

The behavior is a bit different between Windows 10 and Windows 11 as Windows 11 has a different default clipping behavior I believe, so it may work slightly better. Anyway, solid background will seem fine.

For 2) the transition seems a lot smoother now with the image! 🎉

Saw there was another sample for the StartTransitionAction. I spammed the buttons a bit and rarely was able to see a case where it jumped to the finished start/end state (saw it in both directions) without the animation in between. Not sure if a rare timing/threading issue or something else, and not sure if due to transition helpers, behaviors, or animation API/usage itself. Just wanted to call it out though, probably not a big issue, but something we should track. I couldn't see any consistent repro to it. Saw it maybe only once out of every 20-30 transitions? FYI @Sergio0694 in case he has any tips for debugging such things.

Was also wondering @HHChaos if a developer used this on items in a DataTemplate, does that work? I'm imagining a scenario where they have a list of items, and then have a popup/overlay type dialog (that's also contained on their page - as I assume that's a requirement for this API?) that they want to display and how the item coming from? We're close to having labs in a good state to at least start propping up some experiments privately, so we can wait to try this out over there where it may be easier to craft a sample and play with directly, though have to think about how the animation dependency works in Labs. 😋

d2dyno1 commented on Microsoft.Toolkit.Uwp.UI.Animations/Enums/TransitionMode.cs:
```diff
@@ -0,0 +1,22 @@
+        /// <summary>
+        /// The transition strategy for image or other UI elements that require smoother transitions.
+        /// </summary>
+        Image,
```
Can this be named `Smoother`? (Since the above comment implies that it can be used for UI elements as well)

michael-hawker commented:
@HHChaos we want to move this over to Labs. Is that something you want to learn how to do, or want us to help assist with the initial migration (we can still credit the commit to your account)? Let us know, in either case, @niels9001 and I can help out here.

HHChaos commented:
Thanks @michael-hawker @niels9001 for help! It would be best if you could help with the initial migration.

michael-hawker commented:
@HHChaos thanks for continuing to update this PR. 🦙❤ As discussed previously, I've moved this PR to Labs. The new PR is here: https://github.com/CommunityToolkit/Labs-Windows/pull/354

Get better soon, we're all wishing you well.
